R/get_latest_resource_id.R

```diff
@@ -0,0 +1,96 @@
+#' get_latest_resource_id
+#'
+#' to be confident that the resource returned is the one intended
+#' two conditions have to be met. It has to appear at the top of
+#' of the resource list as shown on the open data platform.
+#' The order they are returned via the api is the same
+#' as they appear on the open data platform. It also
+#' has to have the most recent date created
+#'
+#' There are only some datasets that this functionality
+#' is relevant to, these are listed within applicable
+#' datasets and are the datasets that keep historic
+#' resources instead of over writing them.
+#'
+#' @inheritParams get_dataset
+#'
+#' @return a string with the resource id
+get_latest_resource_id <- function(dataset_name) {
+  applicable_datasets <- c(
+    "gp-practice-populations", "gp-practice-contact-details-and-list-sizes",
+    "nhsscotland-payments-to-general-practice", "dental-practices-and-patient-registrations",
+    "general-practitioner-contact-details", "prescribed-dispensed",
+    "prescriptions-in-the-community", "community-pharmacy-contractor-activity"
+  )
+
+  # throw error if name type/format is invalid
+  check_dataset_name(dataset_name)
+
+  # define query and try API call
+  query <- list("id" = dataset_name)
+  content <- try(
+    phs_GET("package_show", query),
+    silent = TRUE
+  )
+
+
+  # check if data set is within applicable datasets
+  # throw error if not
+  if (!dataset_name %in% applicable_datasets) {
+    cli::cli_abort(c(
+      "The dataset name supplied {.var {dataset_name}} is not within the applicable datasets.
+      These are:\n
+      {.var {applicable_datasets}}",
+      "x" = "Please see get_latest_reource documentation.",
+      "i" = "You can find dataset names in the URL
+      of a dataset's page on {.url www.opendata.nhs.scot}."
+    ))
+  }
+
+  # if content contains a 'Not Found Error'
+  # throw error with suggested dataset name
+  if (grepl("Not Found Error", content[1])) {
+    suggest_dataset_name(dataset_name)
+  }
```

Comment:
This code will never be used due to the check above. It should either come before, or is not needed. Alternatively, we could adapt `suggest_dataset_name` to take a custom subset of dataset names e.g.
```r
suggest_dataset_name <- function(dataset_name, dataset_names = "all") {
  if (dataset_names == "all") {
    content <- phs_GET("package_list", "")

    dataset_names <- unlist(content$result)
  }
...
```
Then we could use that to suggest one of the applicable datasets `suggest_dataset_name(dataset_name, applicable_datasets)`

Reply:
thanks, I think it would make sense to put this check after

`  # define query and try API call
  query <- list("id" = dataset_name)
  content <- try(
    phs_GET("package_show", query),
    silent = TRUE
  )
`
?

so that is the user will have more information on why the dataset name was wrong. i.e they will know if the name is incorrect or if the name is correct but is not one of the applicable datasets?

Reply:
Agree with James`s suggestion, I think failing faster for non-applicable datasets would make sense. 
I think as long as there`s only a handful of applicable datasets that are returned with the error message anyway probably no need for suggest_dataset_name as the user should be able to spot & fix the typo by comparing against the returned list of applicable dataset names?

Reply:
ok, thanks. Yeh, I think altering `suggest_dataset_name` would be good. Do you think it an additional argument to trigger this new usage would be the right way to do it? Or create a new function that is used in this context?